test/langtools/jdk/jshell/ToolCompletionTest.java

```diff
@@ -0,0 +1,204 @@
+        String jarName1 = "test1.jar";
+        compiler.jar(dir1, jarName1, "p1/p2/Test.class", "p1/p3/Test.class");
+
+        Path dir2 = outDir.resolve("dir2");
+        compiler.compile(dir2,
+                """
+                package p1.p5;
+                public class Test {
+                }
+                """,
+                """
+                package p1.p6;
+                public class Test {
+                }
+                """);
+        String jarName2 = "test2.jar";
+        compiler.jar(dir2, jarName2, "p1/p5/Test.class", "p1/p6/Test.class");
+
+        test(false, new String[]{"--no-startup", "--class-path", compiler.getPath(dir1.resolve(jarName1)).toString()},
+                (a) -> assertCommand(a, "1", null),
+                (a) -> assertCommand(a, "/env --class-path " + compiler.getPath(dir2.resolve(jarName2)).toString(), null),
+                (a) -> assertCompletions(a, "p1.", ".*p5\\..*p6\\..*"),
+                 //cancel the input, so that JShell can be finished:
+                (a) -> assertCommand(a, "\003", null)
+                );
+    }
+
+    @Test
+    public void testModulePathOnCmdLineIndexing() {
+        Path p1 = outDir.resolve("dir1");
+        compiler.compile(p1,
+                """
+                module m {
+                    exports p1.p2;
+                    exports p1.p3;
+                }
+                """,
+                """
+                package p1.p2;
+                public class Test {
+                }
+                """,
+                """
+                package p1.p3;
+                public class Test {
+                }
+                """);
+        String jarName = "test.jar";
+        compiler.jar(p1, jarName, "p1/p2/Test.class", "p1/p3/Test.class");
+
+        test(false, new String[]{"--no-startup", "--module-path", compiler.getPath(p1.resolve(jarName)).toString()},
+                (a) -> assertCompletions(a, "p1.", ".*p2\\..*p3\\..*"),
+                 //cancel the input, so that JShell can be finished:
+                (a) -> assertCommand(a, "\003", null)
+                );
+    }
+
+    @Test
+    public void testModulePathOnCmdLineIndexing2() throws IOException {
+        Path p1 = outDir.resolve("dir1");
+        compiler.compile(p1,
+                """
+                module m {
+                    exports p1.p2;
+                    exports p1.p3;
+                }
+                """,
+                """
+                package p1.p2;
+                public class Test {
+                }
+                """,
+                """
+                package p1.p3;
+                public class Test {
+                }
+                """);
+        String jarName = "test.jar";
+        Path lib = outDir.resolve("lib");
+        Files.createDirectories(lib);
+        compiler.jar(p1, jarName, "p1/p2/Test.class", "p1/p3/Test.class");
+        Files.move(compiler.getPath(p1.resolve(jarName)), lib.resolve(jarName));
```

Comment:
```suggestion
        Files.copy(compiler.getPath(p1.resolve(jarName)), lib.resolve(jarName));
```

Prevent `java.nio.file.FileSystemException` on Windows:

```
ToolCompletionTest.d\\tool_completion_test\\dir1\\test.jar -> tool_completion_test\\lib\\test.jar: The process cannot access the file because it is being used by another process
```

Seems like `compiler.jar(p1, jarName, "p1/p2/Test.class", "p1/p3/Test.class");` keeps an open file handle to the created `JAR` file.

Reply:
Superseded by https://github.com/openjdk/jdk/pull/26137/commits/54eed125f92aeb016502a8c06b2f284d4e4d4ded